### ISSUE TITLE
Domain Controller fails in LXC

### ISSUE THREAD

**Issue by JedMeister** [turnkeylinux/tracker#1666]
TL;DR: Samba fails in TurnKey Domain Controller. The best path for now is to use a VM instead. To persevere with LXC, please read on.

----

When initialising the [TurnKey Domain Controller](https://www.turnkeylinux.org/domain-controller) on Proxmox LXC, it fails. Here's is part of the stacktrace:

```
[...]
Setting up self join 
Security context active token stack underflow!
PANIC (pid 3299): Security context active token stack underflow!
BACKTRACE: 45 stack frames:
 #0 /lib/x86_64-linux-gnu/libsamba-util.so.0(log_stack_trace +0x32) [0x7f6d959228d2] 
 #1 /lib/x86_64-linux-gnu/libsmbconf.so.0(smb_panic_s3+0x20) [0x7f6d936ce1c0]
 #2 /lib/x86_64-linux-gnu/libsamba-util.so.0(smb_panic+0x2f) [0x7f6d959229df]
 #3 /usr/lib/x86_64-linux-gnu/samba/libsmbd-base.so.0(sec_ct x_active_token+0x72) [0x7f6d9339b1a2]
[...]
```

It appears that it's actually an issue with Samba, rather than LXC (see this closed, "won't fix" [LXC bug report](https://github.com/lxc/lxc/issues/2708)).

There is a [workaround](https://github.com/lxc/lxc/issues/2708#issuecomment-473466062) noted and it appears to work.

I propose that we have a firstboot hook that checks to see if the appliance is running in LXC (`systemd-detect-virt`) and if so apply the workaround.

Having said that, the file that needs to be tweaked (`/usr/share/samba/setup/idmap_init.ldif`) is handled by package management. So we'd also need to update the file for future package upgrades. So we'd need to consider using `icron` or similar to re-apply the tweak if/when the file is updated. It seems that `smb.conf` may also need to be tweaked (to also ensure UIDs of 65533 or less) but I can't yet confirm that.

It's perhaps also worth checking for `xattrs` and directing users on how to do that (also mentioned in that workaround post). It seems using btrfs is a workaround to support that.

**Comment by JedMeister** [comment]
Argh, that still isn't enough...! Now I get:
```
Setting up self join
set_nt_acl_no_snum: fset_nt_acl returned
NT_STATUS_ACCESS_DENIED.
ERROR(runtime): uncaught exception - (3221225506,
'{Access Denied} A process has requested access to an
object but has not been granted those access rights.')

See /var/log/inithooks/samba_dc.log for full output
```
It seems that the only workaround for that is to run in a privileged container (or at least grant `SYS_ADMIN` privilege).

So for now, I'm not sure that we can actually "fix" this and my previous suggestion/post is still not enough...

**Comment by AdalbertKing** [comment]
It's working! -Turnkney Domain Controler in LXC ver 16.2.1 in Proxmox 7.2 and 6.4


Tested under Proxmox host ver 6.4 and 7.2.

**Privilleged LXC.**

 
First prepare  domain controller DC1.net.lan

On start  you have to set in Proxmox GUI in host searched domain (net.lan) , and DNS 8.8.8.8 )
 
Second: I've set DNS and domain during configuring LXC container from WEBGUI of proxmox
 
WIN10 normally joined to domain, and see network folders.
 

If you want to join additional DC2.net.lan, as second domain controller it's works too:

You have to check (by breaking turnkey starting script CTRL+C) in LXC configuration files:

/etc/resolv.conf

/etc/krb5.conf

[https://wiki.samba.org/index.php/Joining_a_Samba_DC_to_an_Existing_Activ...](https://wiki.samba.org/index.php/Joining_a_Samba_DC_to_an_Existing_Active_Directory)

and If you had to correct those configuration files, you had to restart LXC with DC2. After this Joining should go well.

If no, use command on DC2.NET.LAN (after check conf files ):

samba-tool domain join net.lan DC -U"NET\administrator"

LXC features enabled:
Nesting, NFS, SMB/CIFS, Create device Nodes

If you have two nodes Proxmox cluster, you can start first domain controller DC1 on first node of Proxmox pve1, and on second node pve2 additional domain controller DC2.
So, you have finally two domain controllers on two different machines. Both are using all features of ZFS file system  from the proxmox hosts.

**Comment by JedMeister** [comment]
I'm going to close this now. AFAIK running our domain controller in an unprivileged container with nesting enabled appears to allow it to work fine...

**Comment by cryptomilk** [comment]
## How get it working in an unprivileged container?

The problem is the `acl_xattr:security_acl_name` default config option which doesn't work in an unprivileged container (`man vfs_acl_xattr`).

You need to use a different location and be careful with permissions, read the manpage especially the security warnings.

**You need Samba 4.17.0 or newer!**

**Comment by JedMeister** [comment]
@cryptomilk - thanks for the additional info. Following your lead there, google turned this possible workaround up:

https://lists.samba.org/archive/samba/2023-January/243659.html

Although it didn't work cleanly for them:
https://lists.samba.org/archive/samba/2023-January/243660.html

A Samba dev suggests using a newer version of Samba (i.e. from backports):
https://lists.samba.org/archive/samba/2023-January/243661.html

IIRC we already install samba from bullseye-backports, but it likely has a newer version there now.

I hope to get an updated v18.0 release (based on Bookworm) built early next year.

If you discover anything that makes it "just work" in an unpriveledged LXC container, please do share.

**Comment by cryptomilk** [comment]
Hi Jed,

someone from the Samba Team ran into the issue yesterday. We discussed the issue the problem is `acl_xattr:security_acl_name`. The panic still exist if you don't change it and we are looking into it to fix it.

I dunno which Samba version you use but you need Samba 4.17.0 or newer!

```
git tag --contains f2b6258b
samba-4.17.0
```

Do not use vfs_acl_tdb if you have xattr support! If you have to use it, make sure to make daily if there is a lot of file activity make hourly backups of the database!

**Comment by JedMeister** [comment]
@cryptomilk - Apologies for late reply. Not sure how I missed this previously. Anyway, thanks for the additional info, it's muchly appreciated.

> someone from the Samba Team ran into the issue yesterday. We discussed the issue the problem is `acl_xattr:security_acl_name`. The panic still exist if you don't change it and we are looking into it to fix it.

Yeah, that makes sense as AFAIK ACLs aren't supported in an unprivileged container - at least not for ext2/3/4. I vaguely recall reading at some point that ACLs will work in an unprivileged container if the container filesystem is using one of the alternate "more advanced" filesystems supported by Debian? Sorry I don't recall which one, or where or when I read that - so may or may not be the case. Regardless, it might be worth following up on if that's of any interest to you?

FYI, the current TurnKey release (v17.1 - based on Debian 11/Bullseye) [includes samba v4.17.6](https://releases.turnkeylinux.org/turnkey-domain-controller/17.1-bullseye-amd64/turnkey-domain-controller-17.1-bullseye-amd64.manifest) but I note that Samba is upgradeable to [v4.17.12](https://packages.debian.org/bullseye-backports/samba). Note that the Debian bullseye-backports is already enabled and samba (and friends) are [pinned](https://github.com/turnkeylinux-apps/domain-controller/blob/master/Makefile#L5) (that line results in a [apt prefs file equivalent to the one removed here](https://github.com/turnkeylinux-apps/domain-controller/commit/6621ca3226ccce91b781961d85f5d919a6fd6386)) to install from backports. So that newer version can be installed on the current appliance by running `apt update && apt upgrade`.

The v18.x appliance (which hopefully won't be too far away) will include [that same version of samba (v4.17.12)](https://packages.debian.org/bookworm/samba) - but I note that [v4.19.4 is in bookworm backports](https://packages.debian.org/bookworm-backports/samba) so that will be an option if you want/need something even newer. FYI we prefer packages from main where possible as then they get the automated security updates. We moved to the version in backports because it is required to join newer builds of Win 10/11 to a Samba AD domain.

FWIW I downloaded the git repo from [Debian Salsa](https://salsa.debian.org/samba-team/samba/) (the Debian GitLab instance) and checked out the [bookworm branch](https://salsa.debian.org/samba-team/samba/-/tree/bookworm) and I can confirm that every version released in Debian from v4.17.0 onwards includes commit `f2b6258b`.
